src/utils/Render.js

```diff
@@ -54,80 +56,376 @@ function usingDispatchBuffer(buffer, Component) {
 class DispatchBuffer {
   constructor(renderStrategy) {
     this.promisesBuffer = []
-    this.locked = false
+    this.status = {}
+    this.dispatches = []
+    this.promises = []
     this.renderStrategy = renderStrategy
   }
 
-  push(v) {
-    this.promisesBuffer.push(v)
+  push(id, promise) {
+    this.promisesBuffer.push(promise)
+    promise.then(
+      () => this.status[id] = STAT.DONE,
+      () => this.status[id] = STAT.FAILED
+    )
+  }
+
+  getStatus(id) {
+    return this.status[id] || STAT.READY
+  }
+
+  getPromise(f, args) {
+    const res = this.promises.filter((o) => {
+      // same function
+      // same length
+      // every arg shallowEquals itself
+      return o.f === f &&
+        o.args.length === args.length &&
+        o.args.every((arg, i) => shallowEqual(arg, args[i]))
+    })
+
+    return res.length ? res[0].promise : null
   }
 
-  fill(Element) {
-    return this.renderStrategy(Element)
+  savePromise(f, args, promise) {
+    this.promises.push({ f, args, promise })
   }
 
   clear() {
     this.promisesBuffer = []
+    this.promises = []
   }
 
-  flush(alt, Element) {
-    return Promise.all(this.promisesBuffer).then((data) => {
-      // fire off all the actions synchronously
-      data.forEach((f) => {
-        if (Array.isArray(f)) {
-          f.forEach(x => x())
-        } else {
-          f()
-        }
-      })
-      this.locked = true
+  resolve(error, html, alt, Element, i) {
+    return Promise.resolve({
+      error,
+      html,
+      state: alt.flush(),
+      buffer: {
+        status: this.status
+      },
+      element: Element,
+      diagnostics: {
+        iterations: i,
+        dispatches: this.dispatches.length
+      }
+    })
+  }
 
-      return {
-        html: this.renderStrategy(Element),
-        state: alt.flush(),
-        element: Element
+  drain() {
+    // fire off all the actions synchronously
+    this.dispatches.forEach((f) => {
+      if (Array.isArray(f)) {
+        f.forEach(x => x())
+      } else {
+        f()
       }
-    }).catch((err) => {
-      return Promise.reject({
-        err,
-        state: alt.flush(),
-        element: Element
-      })
     })
   }
-}
 
+  render(alt, Element, info) {
+    alt.recycle()
+
+    const startTime = Date.now()
+    const i = info.i || 0
+
+    this.drain()
+
+    // render the html
+    const html = this.renderStrategy(Element)
+
+    if (i >= info.maxIterations) {
+      return this.resolve(
+        new Error('Max number of iterations reached'),
+        html,
+        alt,
+        Element,
+        i
+      )
+    }
+
+    if (info.time > info.timeout) {
```

Comment:
This is probably not the ideal way to handle timeouts - seems like it'd be better to `Promise.race` for the timeout, or have a flag on the `DispatchBuffer` that gets triggered by the timeout firing; otherwise you end up waiting unnecessarily long.